File: packages/sdk-py/src/smscode/resources/catalog.py
from __future__ import annotations

from collections.abc import Awaitable, Callable, Mapping
from typing import Any

from smscode.errors import InvalidResponseError
from smscode.models import (
    Country,
    ExchangeRate,
    Product,
    ProductsPageV1,
    ProductsPageV2,
    ProductV2,
    Service,
    V2Fx,
    parse_v2_fx,
)
from smscode.money import parse_money
from smscode.types import ApiResult, HttpMethod, QueryValue, SortQuery

SyncRequest = Callable[
    [HttpMethod, str],
    ApiResult[Any],
]
AsyncRequest = Callable[
    [HttpMethod, str],
    Awaitable[ApiResult[Any]],
]


def _products_params(
    *,
    country_id: int | None = None,
    platform_id: int | None = None,
    limit: int | None = None,
    page: int | None = None,
    sort: SortQuery | None = None,
) -> dict[str, QueryValue]:
    return {
        "country_id": country_id,
        "platform_id": platform_id,
        "limit": limit,
        "page": page,
        "sort": sort,
    }


def _services_params(*, country_id: int | None = None) -> dict[str, QueryValue]:
    return {"country_id": country_id}


def _exchange_rate_params(*, pair: str | None = None) -> dict[str, QueryValue]:
    return {"pair": pair}


def _decode_v2_products(result: ApiResult[Any]) -> ProductsPageV2:
    meta = result.meta or {}
    fx = parse_v2_fx(meta.get("fx"))
    products: list[Product] = [
        _decode_v2_product(item) for item in result.data if isinstance(item, Mapping)
    ]
    return ProductsPageV2(
        products=products,
        page=_int_or_none(meta.get("page")),
        limit=_int_or_none(meta.get("limit")),
        count=_int_or_none(meta.get("count")),
        fx=fx,
    )


def _decode_v1_products(result: ApiResult[Any]) -> ProductsPageV1:
    meta = result.meta or {}
    products = [_decode_v1_product(item) for item in result.data if isinstance(item, Mapping)]
    return ProductsPageV1(
        products=products,
        page=_int_or_none(meta.get("page")),
        limit=_int_or_none(meta.get("limit")),
        count=_int_or_none(meta.get("count")),
    )


def _decode_v2_product(item: Mapping[str, Any]) -> ProductV2:
    raw = dict(item)
    return ProductV2(
        id=int(item["id"]),
        name=item.get("name") if isinstance(item.get("name"), str) else None,
        country_id=_int_or_none(item.get("country_id")),
        platform_id=_int_or_none(item.get("platform_id")),
        available=int(item["available"]),
        price=parse_money(item.get("price")),
        active=bool(item["active"]),
        catalog_product_id=_int_or_none(item.get("catalog_product_id")),
        raw=raw,
    )


def _decode_v1_product(item: Mapping[str, Any]) -> Product:
    raw = dict(item)
    return Product(
        id=int(item["id"]),
        name=item.get("name") if isinstance(item.get("name"), str) else None,
        country_id=_int_or_none(item.get("country_id")),
        platform_id=_int_or_none(item.get("platform_id")),
        available=int(item["available"]),
        price=item.get("price") if type(item.get("price")) is int else None,
        active=bool(item["active"]),
        catalog_product_id=_int_or_none(item.get("catalog_product_id")),
        raw=raw,
    )


def _decode_country(item: Mapping[str, Any]) -> Country:
    raw = dict(item)
    return Country(
        id=int(item["id"]),
        name=str(item["name"]),
        code=item.get("code") if isinstance(item.get("code"), str) else None,
        dial_code=item.get("dial_code") if isinstance(item.get("dial_code"), str) else None,
        emoji=item.get("emoji") if isinstance(item.get("emoji"), str) else None,
        active=bool(item.get("active", True)),
        raw=raw,
    )


def _decode_service(item: Mapping[str, Any]) -> Service:
    raw = dict(item)
    return Service(
        id=int(item["id"]),
        name=str(item["name"]),
        code=str(item["code"]),
        active=bool(item.get("active", True)),
        raw=raw,
    )


def _decode_countries(data: object) -> list[Country]:
    return (
        [_decode_country(item) for item in data if isinstance(item, Mapping)]
        if isinstance(data, list)
        else []
    )


def _decode_services(data: object) -> list[Service]:
    return (
        [_decode_service(item) for item in data if isinstance(item, Mapping)]
        if isinstance(data, list)
        else []
    )


def _decode_exchange_rate(data: object) -> ExchangeRate:
    if not isinstance(data, Mapping):
        raise InvalidResponseError("The exchange-rate response is malformed.")
    raw = dict(data)
    return ExchangeRate(
        pair=str(data["pair"]),
        rate=data["rate"] if isinstance(data.get("rate"), (int, str)) else str(data.get("rate")),
        base_currency=data.get("base_currency")
        if isinstance(data.get("base_currency"), str)
        else None,
        quote_currency=data.get("quote_currency")
        if isinstance(data.get("quote_currency"), str)
        else None,
        rate_as_of=data.get("rate_as_of") if isinstance(data.get("rate_as_of"), str) else None,
        raw=raw,
    )


def _int_or_none(value: object) -> int | None:
    return value if type(value) is int else None


class V2CatalogResource:
    def __init__(self, request: Any) -> None:
        self._request = request

    def countries(self) -> list[Country]:
        result = self._request("GET", "/v2/catalog/countries")
        return _decode_countries(result.data)

    def services(self, *, country_id: int | None = None) -> list[Service]:
        result = self._request(
            "GET",
            "/v2/catalog/services",
            params=_services_params(country_id=country_id),
        )
        return _decode_services(result.data)

    def products(
        self,
        *,
        country_id: int | None = None,
        platform_id: int | None = None,
        limit: int | None = None,
        page: int | None = None,
        sort: SortQuery | None = None,
    ) -> ProductsPageV2:
        result = self._request(
            "GET",
            "/v2/catalog/products",
            params=_products_params(
                country_id=country_id,
                platform_id=platform_id,
                limit=limit,
                page=page,
                sort=sort,
            ),
        )
        return _decode_v2_products(result)

    def exchange_rate(self, *, pair: str | None = None) -> V2Fx:
        result = self._request(
            "GET",
            "/v2/catalog/exchange-rate",
            params=_exchange_rate_params(pair=pair),
        )
        return parse_v2_fx(result.data)


class V1CatalogResource:
    def __init__(self, request: Any) -> None:
        self._request = request

    def countries(self) -> list[Country]:
        result = self._request("GET", "/v1/catalog/countries")
        return _decode_countries(result.data)

    def services(self, *, country_id: int | None = None) -> list[Service]:
        result = self._request(
            "GET",
            "/v1/catalog/services",
            params=_services_params(country_id=country_id),
        )
        return _decode_services(result.data)

    def products(
        self,
        *,
        country_id: int | None = None,
        platform_id: int | None = None,
        limit: int | None = None,
        page: int | None = None,
        sort: SortQuery | None = None,
    ) -> ProductsPageV1:
        result = self._request(
            "GET",
            "/v1/catalog/products",
            params=_products_params(
                country_id=country_id,
                platform_id=platform_id,
                limit=limit,
                page=page,
                sort=sort,
            ),
        )
        return _decode_v1_products(result)

    def exchange_rate(self, *, pair: str | None = None) -> ExchangeRate:
        result = self._request(
            "GET",
            "/v1/catalog/exchange-rate",
            params=_exchange_rate_params(pair=pair),
        )
        return _decode_exchange_rate(result.data)


class AsyncV2CatalogResource:
    def __init__(self, request: Any) -> None:
        self._request = request

    async def countries(self) -> list[Country]:
        result = await self._request("GET", "/v2/catalog/countries")
        return _decode_countries(result.data)

    async def services(self, *, country_id: int | None = None) -> list[Service]:
        result = await self._request(
            "GET",
            "/v2/catalog/services",
            params=_services_params(country_id=country_id),
        )
        return _decode_services(result.data)

    async def products(
        self,
        *,
        country_id: int | None = None,
        platform_id: int | None = None,
        limit: int | None = None,
        page: int | None = None,
        sort: SortQuery | None = None,
    ) -> ProductsPageV2:
        result = await self._request(
            "GET",
            "/v2/catalog/products",
            params=_products_params(
                country_id=country_id,
                platform_id=platform_id,
                limit=limit,
                page=page,
                sort=sort,
            ),
        )
        return _decode_v2_products(result)

    async def exchange_rate(self, *, pair: str | None = None) -> V2Fx:
        result = await self._request(
            "GET",
            "/v2/catalog/exchange-rate",
            params=_exchange_rate_params(pair=pair),
        )
        return parse_v2_fx(result.data)


class AsyncV1CatalogResource:
    def __init__(self, request: Any) -> None:
        self._request = request

    async def countries(self) -> list[Country]:
        result = await self._request("GET", "/v1/catalog/countries")
        return _decode_countries(result.data)

    async def services(self, *, country_id: int | None = None) -> list[Service]:
        result = await self._request(
            "GET",
            "/v1/catalog/services",
            params=_services_params(country_id=country_id),
        )
        return _decode_services(result.data)

    async def products(
        self,
        *,
        country_id: int | None = None,
        platform_id: int | None = None,
        limit: int | None = None,
        page: int | None = None,
        sort: SortQuery | None = None,
    ) -> ProductsPageV1:
        result = await self._request(
            "GET",
            "/v1/catalog/products",
            params=_products_params(
                country_id=country_id,
                platform_id=platform_id,
                limit=limit,
                page=page,
                sort=sort,
            ),
        )
        return _decode_v1_products(result)

    async def exchange_rate(self, *, pair: str | None = None) -> ExchangeRate:
        result = await self._request(
            "GET",
            "/v1/catalog/exchange-rate",
            params=_exchange_rate_params(pair=pair),
        )
        return _decode_exchange_rate(result.data)

File: packages/sdk-js/src/resources/catalog.ts
/**
 * Catalog read resources — countries, services, products, exchange rate.
 *
 * Two surfaces are exposed by the client:
 * - {@link V2CatalogResource} (the default, `client.catalog.*`) projects prices
 *   to USD {@link Money} objects and carries the per-response FX receipt.
 * - {@link V1CatalogResource} (`client.v1.catalog.*`) returns the canonical IDR
 *   shapes verbatim — `price` is a plain IDR `number`.
 *
 * Both are read-only (`GET`); they delegate to {@link RequestFn} (the client's
 * `request`) and never touch transport details themselves.
 */
import type { ApiResult, QueryValue } from "../client.js";
import { requireFx } from "../internal/decode.js";
import { parseMoney } from "../money.js";
import type { Money } from "../money.js";
import type { components } from "../types.gen.js";

/** The subset of the client used by resources: its `request` method. */
export type RequestFn = <T = unknown>(
  method: "GET" | "POST" | "PUT" | "PATCH" | "DELETE",
  path: string,
  options?: { query?: Record<string, QueryValue> },
) => Promise<ApiResult<T>>;

type V1Country = components["schemas"]["V1Country"];
type V1Service = components["schemas"]["V1Service"];
type V1Product = components["schemas"]["V1Product"];
type V2Product = components["schemas"]["V2Product"];
type V1ExchangeRate = components["schemas"]["V1ExchangeRate"];
type V2Fx = components["schemas"]["V2Fx"];
type PaginationMeta = components["schemas"]["PaginationMeta"];
type V2ProductsMeta = components["schemas"]["V2ProductsMeta"];

/** Query parameters shared by the `services` listing. */
export interface ServicesParams {
  /** Filter by country (`V1Country.id`). */
  country_id?: number;
}

/** Query parameters for the `products` listing. */
export interface ProductsParams {
  /** Filter by country (`V1Country.id`). */
  country_id?: number;
  /** Filter by service/platform (`V1Service.id`). */
  platform_id?: number;
  /** Max items per page (server-clamped 1–10000; default 1000). */
  limit?: number;
  /** 1-based page number (default 1). */
  page?: number;
  /** Sort order for the product list. */
  sort?: components["parameters"]["SortQuery"];
}

/** Query parameters for the exchange-rate lookup. */
export interface ExchangeRateParams {
  /** Currency pair, formatted `BASE/QUOTE` (default `USD/IDR`). */
  pair?: string;
}

/** A decoded `/v2` product — identical to {@link V2Product} but `price` is a {@link Money}. */
export interface ProductV2 extends Omit<V2Product, "price"> {
  price: Money;
}

/** A page of `/v2` products with pagination metadata + the FX receipt. */
export interface ProductsPageV2 {
  products: ProductV2[];
  meta: V2ProductsMeta;
  fx: V2Fx;
}

/** A page of `/v1` products (IDR prices) with pagination metadata. */
export interface ProductsPageV1 {
  products: V1Product[];
  meta: PaginationMeta;
}

/** The `/v2` catalog surface (USD-projected). */
export class V2CatalogResource {
  constructor(private readonly request: RequestFn) {}

  /** List every country in the catalog. */
  async countries(): Promise<V1Country[]> {
    const { data } = await this.request<V1Country[]>(
      "GET",
      "/v2/catalog/countries",
    );
    return data ?? [];
  }

  /** List services, optionally scoped to a country. */
  async services(params: ServicesParams = {}): Promise<V1Service[]> {
    const { data } = await this.request<V1Service[]>(
      "GET",
      "/v2/catalog/services",
      { query: { country_id: params.country_id } },
    );
    return data ?? [];
  }

  /** List a page of products with USD-projected prices + the FX receipt. */
  async products(params: ProductsParams = {}): Promise<ProductsPageV2> {
    const result = await this.request<V2Product[]>("GET", "/v2/catalog/products", {
      query: {
        country_id: params.country_id,
        platform_id: params.platform_id,
        limit: params.limit,
        page: params.page,
        sort: params.sort,
      },
    });
    // Validate the envelope shape with TYPED errors (consistent with orders):
    // `requireFx` throws `INVALID_RESPONSE` on a missing FX receipt (rather than
    // a bare `meta.fx` raw `TypeError` on a malformed 2xx), and `parseMoney`
    // throws `INVALID_MONEY` on a missing/malformed per-product `price`. This is
    // a read; no idempotency key is involved.
    const fx = requireFx(result);
    const meta = result.meta as V2ProductsMeta;
    const products = (result.data ?? []).map((p) => ({
      ...p,
      price: parseMoney(p.price),
    }));
    return { products, meta, fx };
  }

  /** The current USD/IDR FX receipt. */
  async exchangeRate(params: ExchangeRateParams = {}): Promise<V2Fx> {
    const { data } = await this.request<V2Fx>(
      "GET",
      "/v2/catalog/exchange-rate",
      { query: { pair: params.pair } },
    );
    return data as V2Fx;
  }
}

/** The `/v1` catalog surface (canonical IDR). */
export class V1CatalogResource {
  constructor(private readonly request: RequestFn) {}

  /** List every country in the catalog. */
  async countries(): Promise<V1Country[]> {
    const { data } = await this.request<V1Country[]>(
      "GET",
      "/v1/catalog/countries",
    );
    return data ?? [];
  }

  /** List services, optionally scoped to a country. */
  async services(params: ServicesParams = {}): Promise<V1Service[]> {
    const { data } = await this.request<V1Service[]>(
      "GET",
      "/v1/catalog/services",
      { query: { country_id: params.country_id } },
    );
    return data ?? [];
  }

  /** List a page of products with IDR prices + pagination metadata. */
  async products(params: ProductsParams = {}): Promise<ProductsPageV1> {
    const result = await this.request<V1Product[]>("GET", "/v1/catalog/products", {
      query: {
        country_id: params.country_id,
        platform_id: params.platform_id,
        limit: params.limit,
        page: params.page,
        sort: params.sort,
      },
    });
    return {
      products: result.data ?? [],
      meta: result.meta as PaginationMeta,
    };
  }

  /** The stored exchange rate for a currency pair (default `USD/IDR`). */
  async exchangeRate(params: ExchangeRateParams = {}): Promise<V1ExchangeRate> {
    const { data } = await this.request<V1ExchangeRate>(
      "GET",
      "/v1/catalog/exchange-rate",
      { query: { pair: params.pair } },
    );
    return data as V1ExchangeRate;
  }
}
